packages/sync/src/links/OfflineQueueLink.ts

```diff
@@ -62,12 +63,28 @@ export class OfflineQueueLink extends ApolloLink {
     this.operationFilter = filter;
   }
 
-  public open() {
+  public async open() {
     logger("MutationQueue is open", this.opQueue);
     this.isOpen = true;
-    this.opQueue.forEach(({ operation, forward, observer }) => {
-      forward(operation).subscribe(observer);
-    });
+    for (const opEntry of this.opQueue) {
+      const { operation, forward, observer } = opEntry;
+      await new Promise(resolve => {
+        forward(operation).subscribe({
+          next: result => {
+            this.updateIds(opEntry, result);
+            this.storage.setItem(this.key, JSON.stringify(this.opQueue));
```

Comment:
I believe that this is could be the most controversial change. If you have 50 edits in queue for this change we will Parse and Stringify entire queue so many times significantly affecting performance of the application when starting (this will usually happen on startup). To prevent from that we should:

- refactor OfflineRestoreHandler to be part of this so it will be separated from regular processing.
- Avoid setItem being called here as opQueue as we really keep everything in memory and all requests should be processed within this loop.
- Clear out async story - this method is async but we do not do await bellow when calling open etc.

First one can be done after this is merged.

Reply:
The reason for saving the queue back to store is that if the app is closed (or page refreshed) during processing, the changes would not be persisted. The worst thing that could happen is that some updates in queue could later fail because of ids that have not been updated. But I think that if we can easily prevent it we probably should even though the performance would be worse.

Also we do the same currently on master - once there is a response from server, operation is removed from queue and queue is saved to store (`cancelOperation`). Now we do the save twice - when there is a response from server, then when ids are updated. Maybe this could be somehow improved?

Is stringifying that slow? There is still communication with server for every operation, which IMHO will be slower compared to saving queue to store.

Reply:
> The reason for saving the queue back to store is that if the app is closed (or page refreshed) during processing, the changes would not be persisted. 

Hmm... so maybe OfflineRestoreHandler should be triggered by user then at the moment they want to push the changes. I feel like this is really an corner case - if someone refresh page during the sending request to server it will have the same effect.

> The worst thing that could happen is that some updates in queue could later fail because of ids that have not been updated.

After getting thru the changes I think in this case queue will be empty already.